core/src/types.ts

```diff
@@ -266,3 +266,40 @@ export declare interface DebtAuctionEnriched extends DebtAuctionActive {
 }
 
 export declare interface DebtAuctionTransaction extends DebtAuctionEnriched, CompensationAuctionTransactionFees {}
+
+type CollateralType = CollateralConfig['title'];
+export declare interface Collateral {
+    type: CollateralType;
+    stabilityFeeRate: BigNumber;
+    maxDebtPerCollateralUnit: BigNumber;
+    liquidationLimits: LiquidationLimit;
```

Comment:
> There are two liquidation limits that needs to be fetched: global and collateral-based

i was thinking that the global liquidation limit should reside somewhere in store and thus should not be included in the every single vault object.

Reply:
I think it complicates the frontend logic, without a reason: in order to determine `liquidatable` or `not-liquidatable` (which you would otherwise need to do anyway in core) you would need to know those values. I suggest to include both global and collateral limits to the vault object itself. They can be cached to avoid fetching them for every vault

Reply:
When we liquidate one vault , the cache instantly becomes invalid and we can encounter the "limit reached" state: The cached return will claim that we can liquidate a next vault while in real world it's forbidden. We have to keep the limits updated at all times it seems to me. So even if we keep this in core, we still would have to keep the global limit in its separate object to avoid repetition and multiple refetches. On top of that - duplication of information for every vault object seems suboptimal to me.

Reply:
> duplication of information for every vault object seems suboptimal to me.

You don't need to duplicate it (eg store it every time), but enrich vault with those parameters. 

But it's up to implementation, we can also provide limits separately from the vault. But then all limits, as I don't see the reason why collateral-related limits are "duplicated" in every vault, while global are not – both collateral and global limits will be updated as soon as the vault is liquidated. Making two different update mechanisms for this is more than suboptimal. Removing limits cache on `liquidateVault` call and refetching `getVaults` should be sufficient. Implementing complicated refetching limits logic twice (on frontend and in the bot) is sub-optimal in my opinion

Reply:
Ok, then i think the best option is to store them separately per collateral and for global. See the types.

Reply:
How does it solves `Implementing complicated refetching limits logic twice`? Why do we need to fetch limits separately and compare them with vault amounts manually twice in the bot and the frontend? Or how do you imagine it?

Reply:
> Why do we need to fetch limits separately

if the q is actually "why do we need to fetch limits into separate object", then the answer is - we then have a central place and a single source of truth where the limits can be looked up for every vault (collateral type).

If the q is indeed "Why do we need to fetch limits separately", then the answer is - because the contract stores them in separate places and we have to fetch one limit object per collateral + limit object for global scope.

___

we don't need to compare them in frontend or bot, since we have the states `liquidatable` and `not-liquidatable` - we are forced to adjust the state along with the refreshed limits when the vault gets liquidated. This happens in the core. I don't see where it happens twice here.

So fetching vaults triggers the mechanism, core logic sets every vault to have the corresponding state; Then after liquidating each vault we have to update the limits and state because otherwise the data will be invalid. For frontend and bot only `Vaults` object will be relevant while `Limits` will serve to determine the states under the hood and will not be exposed to frontend/bot.

Reply:
> we are forced to adjust the state along with the refreshed limits when the vault gets liquidated. This happens in the core. I don't see where it happens twice here

- Invalidation of the cache will need to be implemented twice on frontend and bot, since you expect them to refetch values after the liquidation
- `Limits` will need to be fetched separately from the `vaults` array twice
  - As they will still need to be provided to the frontend to be displayed as separate numbers (see the latest mock) and will be required for the logs of the bot
- They will need to be transformed twice from `LiquidationLimits.collateralDebts` to another format which doesn't send a record of all collateral limits

I still don't understand why do you need to first _build up a cache by hand_, then use it to enrich the vault with the state (and still provide those two interdependent things separately), when you can just call the function and _cache the results using library_ which will do all the heavy lifting for you?

I still advice to flatten `LiquidationLimits` and merge them into `Vault`